waitUntilCondition method throws KubernetesClientException after approximately one hour with sufficiently long timeout

### Describe the bug

I encountered an issue while using the waitUntilCondition method from the Fabric8 Kubernetes Client library. 

code：
client.pods().inNamespace(namespace).withName(pod.getMetadata().getName()).waitUntilCondition(
                    o -> o.getStatus().getPhase().equals(PodConstant.POD_FAILED_PHASE) || o.getStatus().getPhase().equals(PodConstant.POD_SUCCEEDED_PHASE), timeout, TimeUnit.SECONDS);

However, I observed that when the method is used with a timeout value that is long enough (approximately one hour or more), it throws a KubernetesClientException even though the condition has not been met. This behavior is unexpected and prevents the method from working as intended.

my k8s version is v1.19

### Fabric8 Kubernetes Client version

5.12.4

### Steps to reproduce

Set up a scenario where a Pod's status will change over a longer period of time.
Use the waitUntilCondition method with a timeout value of approximately one hour or more.
Observe that the method throws a KubernetesClientException after the timeout period, even if the condition should be met.

### Expected behavior

The waitUntilCondition method should wait for the specified condition to be met within the provided timeout period. If the condition is met, the method should not throw an exception. If the condition is not met within the timeout period, then the method can throw an exception indicating that the condition was not met.

### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

other (please specify in additional context)

### Environment

Linux

### Fabric8 Kubernetes Client Logs

```shell
io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.

at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:64)

at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:53)

at io.fabric8.kubernetes.client.dsl.base.BaseOperation.waitUntilConditionWithRetries(BaseOperation.java:1149)

at io.fabric8.kubernetes.client.dsl.base.BaseOperation.waitUntilCondition(BaseOperation.java:1115)

at io.fabric8.kubernetes.client.dsl.base.BaseOperation.waitUntilCondition(BaseOperation.java:84)

at com.pinlandata.project.scheduler.executor.util.KubeOperator.waitPodFinished(KubeOperator.java:74)

at com.pinlandata.project.scheduler.executor.util.KubeOperator$$FastClassBySpringCGLIB$$8ec6f2dd.invoke(<generated>)

at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)

at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:779)

at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)

at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)

at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:166)

at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)

at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)

at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:692)

at com.pinlandata.project.scheduler.executor.util.KubeOperator$$EnhancerBySpringCGLIB$$d9f7fe0.waitPodFinished(<generated>)

at com.pinlandata.project.scheduler.executor.util.PodRunner.runningPod(PodRunner.java:91)

at com.pinlandata.project.scheduler.executor.util.PodRunner.run(PodRunner.java:53)

at com.pinlandata.project.scheduler.executor.service.impl.algo.model.ModelInitService.execute(ModelInitService.java:79)

at com.pinlandata.project.scheduler.transporter.service.impl.NormalTaskConsumeService.consume(NormalTaskConsumeService.java:51)

at com.pinlandata.project.scheduler.transporter.mq.impl.NormalTaskListener.receive(NormalTaskListener.java:57)

at sun.reflect.GeneratedMethodAccessor323.invoke(Unknown Source)

at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

at java.lang.reflect.Method.invoke(Method.java:498)

at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171)

at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120)

at org.springframework.amqp.rabbit.listener.adapter.HandlerAdapter.invoke(HandlerAdapter.java:53)

at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:222)

at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.invokeHandlerAndProcessResult(MessagingMessageListenerAdapter.java:150)

at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.onMessage(MessagingMessageListenerAdapter.java:135)

at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:1632)

at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.actualInvokeListener(AbstractMessageListenerContainer.java:1551)

at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:1539)

at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:1530)

at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.executeListener(AbstractMessageListenerContainer.java:1474)

at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.doReceiveAndExecute(SimpleMessageListenerContainer.java:967)

at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.receiveAndExecute(SimpleMessageListenerContainer.java:913)

at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.access$1600(SimpleMessageListenerContainer.java:83)

at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.mainLoop(SimpleMessageListenerContainer.java:1288)

at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.run(SimpleMessageListenerContainer.java:1194)

at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

at java.lang.Thread.run(Thread.java:748)

Caused by: io.fabric8.kubernetes.client.dsl.base.WaitForConditionWatcher$WatchException: Watcher closed

at io.fabric8.kubernetes.client.dsl.base.WaitForConditionWatcher.onClose(WaitForConditionWatcher.java:62)

at io.fabric8.kubernetes.client.utils.WatcherToggle.onClose(WatcherToggle.java:56)

at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.closeEvent(WatchConnectionManager.java:367)

at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.access$700(WatchConnectionManager.java:50)

at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onMessage(WatchConnectionManager.java:258)

at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:322)

at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:219)

at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:105)

at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:273)

at okhttp3.internal.ws.RealWebSocket$1.onResponse(RealWebSocket.java:209)

at okhttp3.RealCall$AsyncCall.execute(RealCall.java:174)

at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)

... 3 common frames omitted

Caused by: io.fabric8.kubernetes.client.KubernetesClientException: too old resource version: 664650743 (664670130)

... 11 common frames omitted
```


### Additional context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 * Fix #5382: [java-generator] Allow to deserialize more valid RFC3339 date-time and make the format customizable
 * Fix #5380: [java-generator] Avoid to emit Java Keywords as package names
 * Fix #5457: [java-generator] Correctly handle numeric enums
+* Fix #5379: ensuring informOnCondition and waitUntilCondition commple with exception when the client closes
 * Fix #5463: ensures that onStopLeading is called with releaseOnCancel even when leadership is already lost
 * Fix #5423: OkHttpClientImpl supports setting request method for empty payload requests
 

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/BaseOperation.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/BaseOperation.java
@@ -990,6 +990,13 @@ public class BaseOperation<T extends HasMetadata, L extends KubernetesResourceLi
         future.completeExceptionally(t);
       }
     });
+    informer.stopped().whenComplete((v, t) -> {
+      if (t != null) {
+        future.completeExceptionally(t);
+      } else {
+        future.completeExceptionally(new KubernetesClientException("Informer was stopped"));
+      }
+    });
     return future;
   }
 

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java
@@ -33,8 +33,6 @@ import io.fabric8.kubernetes.client.dsl.NamespaceListVisitFromServerGetDeleteRec
 import io.fabric8.kubernetes.client.dsl.NamespaceableResource;
 import io.fabric8.kubernetes.client.dsl.Resource;
 import io.fabric8.kubernetes.client.dsl.Waitable;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -55,8 +53,6 @@ public class NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImp
     implements NamespaceListVisitFromServerGetDeleteRecreateWaitApplicable<HasMetadata>,
     Waitable<List<HasMetadata>, HasMetadata> {
 
-  private static final Logger LOGGER = LoggerFactory
-      .getLogger(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.class);
   protected static final String EXPRESSION = "expression";
 
   private OperationContext context;
@@ -133,24 +129,26 @@ public class NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImp
 
     long finish = System.nanoTime() + timeUnit.toNanos(amount);
 
-    for (int i = 0; i < items.size(); i++) {
-      final HasMetadata meta = items.get(i);
-      CompletableFuture<List<HasMetadata>> future = futures.get(i);
-      try {
-        results.add(future.thenApply(l -> l.isEmpty() ? null : l.get(0)).get(Math.max(0, finish - System.nanoTime()),
-            TimeUnit.NANOSECONDS));
-      } catch (TimeoutException e) {
-        itemsWithConditionNotMatched.add(meta);
-        logAsNotReady(e, meta);
-      } catch (ExecutionException e) {
-        itemsWithConditionNotMatched.add(meta);
-        logAsNotReady(e.getCause(), meta);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt();
-        throw KubernetesClientException.launderThrowable(e);
-      } finally {
-        future.cancel(true);
+    try {
+      for (int i = 0; i < items.size(); i++) {
+        final HasMetadata meta = items.get(i);
+        CompletableFuture<List<HasMetadata>> future = futures.get(i);
+        try {
+          results.add(future.thenApply(l -> l.isEmpty() ? null : l.get(0)).get(Math.max(0, finish - System.nanoTime()),
+              TimeUnit.NANOSECONDS));
+        } catch (TimeoutException e) {
+          itemsWithConditionNotMatched.add(meta);
+        } catch (ExecutionException e) {
+          throw KubernetesClientException.launderThrowable(e.getCause());
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw KubernetesClientException.launderThrowable(e);
+        } finally {
+          future.cancel(true);
+        }
       }
+    } finally {
+      futures.forEach(f -> f.cancel(true));
     }
 
     if (!itemsWithConditionNotMatched.isEmpty()) {
@@ -158,13 +156,6 @@ public class NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImp
     }
 
     return results;
-  }
-
-  private static void logAsNotReady(Throwable t, HasMetadata meta) {
-    LOGGER.warn(
-        "Error while waiting for: [{}] with name: [{}] in namespace: [{}]: {}. The resource will be considered not ready.",
-        meta.getKind(), meta.getMetadata().getName(), meta.getMetadata().getNamespace(), t.getMessage());
-    LOGGER.debug("The error stack trace:", t);
   }
 
   @Override

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/ResourceTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/ResourceTest.java
@@ -39,6 +39,7 @@ import io.fabric8.kubernetes.client.readiness.Readiness;
 import io.fabric8.kubernetes.client.server.mock.EnableKubernetesMockClient;
 import io.fabric8.kubernetes.client.server.mock.KubernetesMockServer;
 import io.fabric8.kubernetes.client.utils.Serialization;
+import io.fabric8.kubernetes.client.utils.Utils;
 import okhttp3.mockwebserver.RecordedRequest;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.DisplayName;
@@ -47,6 +48,7 @@ import org.junit.jupiter.api.Test;
 import java.net.HttpURLConnection;
 import java.util.Objects;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 
 import static java.net.HttpURLConnection.HTTP_GONE;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
@@ -445,6 +447,42 @@ class ResourceTest {
     assertThat(p.getStatus().getConditions())
         .extracting("type", "status")
         .containsExactly(tuple("Ready", "True"), tuple("Dummy", "True"));
+  }
+
+  @Test
+  void testWaitUntilConditionClosedClient() throws InterruptedException {
+    Pod pod1 = new PodBuilder().withNewMetadata()
+        .withName("pod1")
+        .withResourceVersion("1")
+        .withNamespace("test")
+        .and()
+        .build();
+
+    Pod noReady = createReadyFrom(pod1, "False", "1");
+    list(noReady);
+
+    server.expect()
+        .get()
+        .withPath(
+            "/api/v1/namespaces/test/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dpod1&resourceVersion=1&timeoutSeconds=600&watch=true")
+        .andUpgradeToWebSocket()
+        .open()
+        .waitFor(4000)
+        .andEmit(new WatchEvent(noReady, "DELETED"))
+        .done()
+        .always();
+
+    Utils.schedule(Runnable::run, client::close, 2, TimeUnit.SECONDS);
+
+    PodResource podResource = client.pods().withName("pod1");
+    KubernetesClientException exception = assertThrows(KubernetesClientException.class, () -> podResource.waitUntilCondition(
+        r -> r.getStatus()
+            .getConditions()
+            .stream()
+            .anyMatch(c -> "Dummy".equals(c.getType()) && "True".equals(c.getStatus())),
+        10, SECONDS));
+    assertEquals("Informer was stopped", exception.getMessage());
+
   }
 
   @Test

